In [1]:
import warnings

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

warnings.filterwarnings("ignore")


%matplotlib inline

In [2]:
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

## 1. 問題の定義

### 解くべき問題は何か？
- ある腫瘍のデータを見て、悪性(Malignant)か良性(Benign)を区別すると

### 現状の予測手法はあるか？
- 専門家(医者)が診断をしている
    
### どういう問題として処理するか？
- 教師あり学習
- 2クラス分類問題


## 2. データの取得・観察


In [3]:
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

In [4]:
# データセットの説明を確認
print(cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean Radius, f

In [5]:
# 扱いやすくするためデータフレームに変換する
X = pd.DataFrame(cancer.data, columns=cancer.feature_names)

y = pd.Series(cancer.target)

In [6]:
# 1:悪性(Malignant)  / 0: 良性(Benign)
y.value_counts()

1    357
0    212
dtype: int64

In [7]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
mean radius                569 non-null float64
mean texture               569 non-null float64
mean perimeter             569 non-null float64
mean area                  569 non-null float64
mean smoothness            569 non-null float64
mean compactness           569 non-null float64
mean concavity             569 non-null float64
mean concave points        569 non-null float64
mean symmetry              569 non-null float64
mean fractal dimension     569 non-null float64
radius error               569 non-null float64
texture error              569 non-null float64
perimeter error            569 non-null float64
area error                 569 non-null float64
smoothness error           569 non-null float64
compactness error          569 non-null float64
concavity error            569 non-null float64
concave points error       569 non-null float64
symmetry error             569 

### データをざっと見て
- 569行のレコード
- 全ての値がfloat
- 欠損値はない

In [8]:
#  それぞれのカラムにおける統計量をある程度の確認することもできる
X.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

## 3. 評価指標と評価プロトコル

### 評価指標
- クラスのバランスが不均衡(212 - Malignant, 357 - Benign)
- 間違いの重要度が異なる(Malignantを陽性とすると)
    - 良性を悪性と診断してしまう(偽陰性)
    - 悪性を良性と診断してしまう(偽陽性)
- F1-scoreを使ってみよう    


### 評価プロトコル
- データの量が多いわけではないので交差検証で行うべきだろう

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [10]:
# 分割後の形状確認(学習データ)
X_train.shape, y_train.shape

((426, 30), (426,))

In [11]:
# 分割後の形状確認(テストデータ)
X_test.shape, y_test.shape

((143, 30), (143,))

## 4. データの前処理・特徴量エンジニアリング

### データの観察と研究
- 今回は割愛
- ひとまずmodelを作って評価してみよう

## 5. モデルの訓練・評価

### モデルの訓練 と モデルの評価
- 交差検証でモデルと評価を行う
- 機械学習アルゴリズムはいくつか試す    

In [12]:
# 交差検証のための関数をimport
from sklearn.model_selection import cross_val_score

### その1: KNN

In [13]:
from sklearn.neighbors import KNeighborsClassifier

cv_score_knn = cross_val_score(KNeighborsClassifier(), X_train, y_train, cv=5)

print(f"Avg score: {np.mean(cv_score_knn):.3f}")
print(f"Std: {cv_score_knn.std():.3f}")

Avg score: 0.920
Std: 0.043


###  その2: LogisticRegression

In [14]:
from sklearn.linear_model import LogisticRegression

cv_score_logreg = cross_val_score(LogisticRegression(), X_train, y_train, cv=5)


print(f"Avg score: {np.mean(cv_score_logreg):.3f}")
print(f"Std: {cv_score_logreg.std():.3f}")

Avg score: 0.948
Std: 0.027


### その3: SupportVectorMachine(スケーリングなし)

In [15]:
from sklearn.svm import SVC

cv_score_svm = cross_val_score(SVC(), X_train, y_train, cv=5)


print(f"Avg score: {np.mean(cv_score_svm):.3f}")
print(f"Std: {cv_score_svm.std():.3f}")

Avg score: 0.629
Std: 0.001


### その4: SVM(スケーリングあり)
- SVMはスケーリングされていないデータに弱いことを思い出す！

In [16]:
# スケーリングをしてみる
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)

cv_score_svm_scaled = cross_val_score(SVC(), X_train_scaled, y_train, cv=5)

print(f"Avg score: {np.mean(cv_score_svm_scaled):.3f}")
print(f"Std: {cv_score_svm_scaled.std():.3f}")

Avg score: 0.939
Std: 0.012


### その5:  KNN(スケーリングあり)

In [17]:
cv_score_knn_scaled = cross_val_score(
    KNeighborsClassifier(), X_train_scaled, y_train, cv=5)

print(f"Avg score: {np.mean(cv_score_knn_scaled):.3f}")
print(f"Std: {cv_score_knn_scaled.std():.3f}")

Avg score: 0.953
Std: 0.017


### その6:  LogisticRegression(スケーリングあり)

In [18]:
cv_score_logreg_scaled = cross_val_score(
    LogisticRegression(), X_train_scaled, y_train, cv=5)


print(f"Avg score: {np.mean(cv_score_logreg_scaled):.3f}")
print(f"Std: {cv_score_logreg_scaled.std():.3f}")

Avg score: 0.953
Std: 0.015


### それぞれの結果を並べてみる


In [19]:
print(f"Avg score(KNN): {np.mean(cv_score_knn):.3f}")
print(f"Avg score(KNN_scaled): {np.mean(cv_score_knn_scaled):.3f}")

print(f"Avg score(Logreg): {np.mean(cv_score_logreg):.3f}")
print(f"Avg score(Logreg_scaled): {np.mean(cv_score_logreg_scaled):.3f}")

print(f"Avg score(SVM): {np.mean(cv_score_svm):.3f}")
print(f"Avg score(SVM_scaled): {np.mean(cv_score_svm_scaled):.3f}")

Avg score(KNN): 0.920
Avg score(KNN_scaled): 0.953
Avg score(Logreg): 0.948
Avg score(Logreg_scaled): 0.953
Avg score(SVM): 0.629
Avg score(SVM_scaled): 0.939


## チューニングその1: Logreg(スケーリングなし)

In [20]:
from sklearn.model_selection import GridSearchCV

param_grid = {"C": [0.001, 0.01, 0.1, 1, 10, 100]}

grid_logreg = GridSearchCV(LogisticRegression(), param_grid, cv=5)

grid_logreg.fit(X_train, y_train)

print(f"{grid_logreg.best_score_:.3f}")
print(grid_logreg.best_params_)

0.962
{'C': 100}


/Users/mohira/src/github.com/mohira/ml-materials/.venv/lib/python3.7/site-packages/sklearn/model_selection/_search.py:813: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


## チューニングその2:  SVM(スケーリングあり)

In [21]:
from sklearn.model_selection import GridSearchCV

param_grid = {"C": [0.001, 0.01, 0.1, 1, 10, 100],
              "gamma": [0.001, 0.01, 0.1, 1, 10, 100]}

grid_svm = GridSearchCV(SVC(), param_grid, cv=5)

grid_svm.fit(X_train_scaled, y_train)

print(f"{grid_svm.best_score_:.3f}")
print(grid_svm.best_params_)

0.979
{'C': 1, 'gamma': 1}


###  混同行列をみてみる

In [22]:
from sklearn.metrics import confusion_matrix

pred_grid_logreg = grid_logreg.predict(X_test)
confusion_matrix(y_test, pred_grid_logreg)

array([[52,  2],
       [ 3, 86]])

In [23]:
X_test_scaled = scaler.transform(X_test)

pred_grid_svm = grid_svm.predict(X_test_scaled)
confusion_matrix(y_test, pred_grid_svm)

array([[52,  2],
       [ 1, 88]])

### F値で評価してみよう

In [24]:
from sklearn.metrics import f1_score

print(f"LogisticRegression: {f1_score(y_test, pred_grid_logreg):.3f}")
print(f"SurpportVectorMachine: {f1_score(y_test, pred_grid_svm):.3f}")

LogisticRegression: 0.972
SurpportVectorMachine: 0.983


### テストデータの最終的な評価

In [25]:
# SVMはスケーリングしていたのでテストデータもスケーリングするのを忘れずに！
X_test_scaled = scaler.transform(X_test)

print(f"{grid_svm.score(X_test_scaled, y_test):.3f}")

0.979
